gateway/ln-gateway/src/lnd.rs

```diff
@@ -0,0 +1,302 @@
+                tonic::Code::Internal,
+                "LND did not return a preimage",
+            )));
+        };
+
+        Ok(PayInvoiceResponse {
+            preimage: send_response.payment_preimage,
+        })
+    }
+
+    async fn subscribe_htlcs<'a>(
+        &self,
+        subscription: SubscribeInterceptHtlcsRequest,
+    ) -> crate::Result<HtlcStream<'a>> {
+        let channel_size = 1024;
+
+        // Channel to send responses to LND after processing intercepted HTLC
+        let (lnd_tx, lnd_rx) = mpsc::channel::<ForwardHtlcInterceptResponse>(channel_size);
+
+        // Channel to send intercepted htlc to gatewayd for processing
+        let (gwd_tx, gwd_rx) =
+            mpsc::channel::<Result<SubscribeInterceptHtlcsResponse, tonic::Status>>(channel_size);
+
+        let scid = subscription.short_channel_id;
+        let mut client = self.client.clone();
+
+        let mut tg = self.task_group.clone();
+        let outcomes = self.outcomes.clone();
+
+        // Spawn an LND interceptor task
+        tg.spawn("LND HTLC Subscription", move |_handle| async move {
+            let mut htlc_stream = match client
+                .router()
+                .htlc_interceptor(ReceiverStream::new(lnd_rx))
+                .await
+                .map_err(|e| {
+                    error!("Failed to connect to lnrpc server: {:?}", e);
+                    GatewayError::Other(anyhow!("Failed to subscribe to LND htlc stream"))
+                }) {
+                Ok(stream) => stream.into_inner(),
+                Err(e) => {
+                    let _ = gwd_tx
+                        .send(Err(tonic::Status::new(
+                            tonic::Code::Internal,
+                            e.to_string(),
+                        )))
+                        .await;
+                    return;
+                }
+            };
+
+            while let Some(htlc) = match htlc_stream.message().await {
+                Ok(htlc) => htlc,
+                Err(e) => {
+                    error!("Error received over HTLC subscriprion: {:?}", e);
+                    let _ = gwd_tx
+                        .send(Err(tonic::Status::new(
+                            tonic::Code::Internal,
+                            e.to_string(),
+                        )))
+                        .await;
+                    None
+                }
+            } {
+                trace!("handling htlc {:?}", htlc);
+                let response: Option<ForwardHtlcInterceptResponse> =
+                    if htlc.outgoing_requested_chan_id != scid {
+                        // Pass through: This HTLC doesn't belong to the current subscription
+                        // Forward it to the next interceptor or next node
+                        Some(ForwardHtlcInterceptResponse {
+                            incoming_circuit_key: htlc.incoming_circuit_key,
+                            action: 2,
```

Comment:
Would prefer if we can use enums here since 2 is difficult to understand what that means

Reply:
great point